### PR TITLE
No-code schema Unit tests #9530

### PR DESCRIPTION
**Pull request by vbradnitski**

### CODE REVIEW

File: modules/lib/lib-app/src/main/resources/lib/xp/app.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
+var value = params[name]/**
```
Comment (alansemenov) [inline modules/lib/lib-app/src/main/resources/lib/xp/app.js:1]: Codacy has a fix for the issue: [Missing semicolon.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
var value = params[name];/**
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createPage.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                  </regions>
+                </page>
+                `
+
+// BEGIN
+// Create virtual page.
+var result = schemaLib.createComponent({
+    key: 'myapp:mypage',
+    type: 'PAGE',
+    resource: resource
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createPage.js:32]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createStyles.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+               '<filter>pixelate(10)</filter>' +
+               '</image>' +
+               '</styles>'
+
+
+// BEGIN
+// Create virtual styles.
+var result = schemaLib.createStyles({
+    key: 'myapp',
+    resource: resource
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createStyles.js:26]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/getContentType.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+});
+
+log.info('Fetched content type: ' + result.name);
+
+// END
+
+
+assert.assertJsonEquals({
+    name: "myapp:mytype",
+    displayName: "My type display name",
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/getContentType.js:20]: Codacy has a fix for the issue: [Strings must use singlequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    displayName: 'My type display name',
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/getContentType.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    description: "My type description",
+    modifiedTime: "2010-01-01T10:00:00Z",
+    resource: "<content-type><some-data></some-data></content-type>",
+    type: "CONTENT_TYPE"
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/getContentType.js:24]: Codacy has a fix for the issue: [Strings must use singlequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    type: 'CONTENT_TYPE'
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createXData.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+            </input>
+          </form>
+        </x-data>`
+
+// BEGIN
+// Create virtual mixin.
+var result = schemaLib.createSchema({
+    name: 'myapp:mydata',
+    type: 'XDATA',
+    resource: resource
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createXData.js:27]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createMixin.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            <mixin name="myapplication:inline"/>
+          </form>
+        </mixin>`
+
+// BEGIN
+// Create virtual mixin.
+var result = schemaLib.createSchema({
+    name: 'myapp:mytype',
+    type: 'MIXIN',
+    resource: resource
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createMixin.js:24]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/updateContentType.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+                </input>
+            </form>
+        </content-type>`
+
+// BEGIN
+// Update virtual content type.
+var result = schemaLib.updateSchema({
+    name: 'myapp:mytype',
+    type: 'CONTENT_TYPE',
+    resource: resource
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/updateContentType.js:34]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/getContentType.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+log.info('Fetched content type: ' + result.name);
+
+// END
+
+
+assert.assertJsonEquals({
+    name: "myapp:mytype",
+    displayName: "My type display name",
+    description: "My type description",
+    modifiedTime: "2010-01-01T10:00:00Z",
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/getContentType.js:22]: Codacy has a fix for the issue: [Strings must use singlequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    modifiedTime: '2010-01-01T10:00:00Z',
```

File: modules/lib/lib-app/src/main/resources/lib/xp/examples/app/list.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
+var appLib = require('/lib/xp/app');
+var assert = require('/lib/xp/testing');
+
+/* global log*/
+
+// BEGIN
+// List virtual apps.
+var result = appLib.list();
+
+log.info('Listed apps: ' + result.map(app => app.key).join(', '));
```
Comment (alansemenov) [inline modules/lib/lib-app/src/main/resources/lib/xp/examples/app/list.js:10]: Codacy has a fix for the issue: [Expected parentheses around arrow function argument.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
log.info('Listed apps: ' + result.map((app) => app.key).join(', '));
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/getContentType.js (hunk 4 above)
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/getContentType.js:20]: Codacy has a fix for the issue: [Strings must use singlequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    displayName: 'My type display name',
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/getContentType.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    type: 'CONTENT_TYPE'
+});
+
+log.info('Fetched content type: ' + result.name);
+
+// END
+
+
+assert.assertJsonEquals({
+    name: "myapp:mytype",
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/getContentType.js:19]: Codacy has a fix for the issue: [Strings must use singlequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    name: 'myapp:mytype',
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createStyles.js [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+               '<filter>pixelate(10)</filter>' +
+               '</image>' +
+               '</styles>'
+
+
+// BEGIN
+// Create virtual styles.
+var result = schemaLib.createStyles({
+    application: 'myapp',
+    resource: resource
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createStyles.js:26]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/test/resources/test/UpdateDynamicContentSchemaHandlerTest.js [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+        <content-type xmlns="urn:enonic:xp:model:1.0">
+        </content-type>`
+    }));
+}
+
+exports.updateInvalidContentSchemaType = function () {
+    assert.assertThrows(() => schemaLib.updateSchema({
+        name: 'myapp:mydata',
+        type: 'INVALID_TYPE',
+        resource: ""
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/test/resources/test/UpdateDynamicContentSchemaHandlerTest.js:20]: Codacy has a fix for the issue: [Strings must use singlequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
        resource: ''
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/getContentType.js [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    displayName: "My type display name",
+    description: "My type description",
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/getContentType.js:21]: Codacy has a fix for the issue: [Strings must use singlequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    description: 'My type description',
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createXData.js (hunk 6 above)
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createXData.js:27]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/getContentType.js (hunk 11 above)
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/getContentType.js:19]: Codacy has a fix for the issue: [Strings must use singlequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    name: 'myapp:mytype',
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/updateXData.js [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+            </input>
+          </form>
+        </x-data>`
+
+// BEGIN
+// Update virtual mixin.
+var result = schemaLib.updateSchema({
+    name: 'myapp:mydata',
+    type: 'XDATA',
+    resource: resource
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/updateXData.js:27]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createPart.js [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+                  </config>
+                </part>
+                `
+
+// BEGIN
+// Create virtual part.
+var result = schemaLib.createComponent({
+    key: 'myapp:mypart',
+    type: 'PART',
+    resource: resource
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createPart.js:39]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/getContentType.js (hunk 4 above)
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/getContentType.js:20]: Codacy has a fix for the issue: [Strings must use singlequote.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    displayName: 'My type display name',
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createMixin.js (hunk 7 above)
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createMixin.js:24]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/updatePart.js [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+                  </config>
+                </part>
+                `
+
+// BEGIN
+// Update virtual part.
+var result = schemaLib.updateComponent({
+    key: 'myapp:mypart',
+    type: 'PART',
+    resource: resource
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/updatePart.js:39]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createLayout.js [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+                  </regions>
+                </layout>
+                `
+
+// BEGIN
+// Create virtual layout.
+var result = schemaLib.createComponent({
+    key: 'myapp:mylayout',
+    type: 'LAYOUT',
+    resource: resource
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createLayout.js:63]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/updatePage.js [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+                  </regions>
+                </page>
+                `
+
+// BEGIN
+// Update virtual page.
+var result = schemaLib.updateComponent({
+    key: 'myapp:mypage',
+    type: 'PAGE',
+    resource: resource
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/updatePage.js:32]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createXData.js (hunk 6 above)
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createXData.js:27]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createMixin.js (hunk 7 above)
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createMixin.js:24]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/updateLayout.js [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+                  </regions>
+                </layout>
+                `
+
+// BEGIN
+// Update virtual layout.
+var result = schemaLib.updateComponent({
+    key: 'myapp:mylayout',
+    type: 'LAYOUT',
+    resource: resource
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/updateLayout.js:63]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/updateStyles.js [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+               '<filter>pixelate(10)</filter>' +
+               '</image>' +
+               '</styles>'
+
+
+// BEGIN
+// Update virtual styles.
+var result = schemaLib.updateStyles({
+    application: 'myapp',
+    resource: resource
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/updateStyles.js:26]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createStyles.js (hunk 12 above)
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/createStyles.js:26]: Codacy has a fix for the issue: [Expected property shorthand.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
    resource
```

File: modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/listParts.js [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+/* global log*/
+
+// BEGIN
+// Fetch virtual parts.
+var result = schemaLib.listComponents({
+    application: 'myapp',
+    type: 'PART'
+});
+
+log.info('Fetched parts: ' + result.map(part => part.key).join(','));
```
Comment (alansemenov) [inline modules/lib/lib-schema/src/main/resources/lib/xp/examples/schema/listParts.js:13]: Codacy has a fix for the issue: [Expected parentheses around arrow function argument.](https://app.codacy.com/gh/enonic/xp/pullRequest?prid=9736743)
```suggestion
log.info('Fetched parts: ' + result.map((part) => part.key).join(','));
```

**Review by rymsha — APPROVED** [review]

**Comment by rymsha** [comment]
Lets' merge it and review the entire epic

**Pull request merged**